### PR TITLE
Fix ui_api_request_limiting tests

### LINKED ISSUE #1882: Fix should_return_too_many_requests_status_using_default_server_max_active_requests  flaky test
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1880#issuecomment-1625480301

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #1882

### CODE REVIEW

File: test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs [hunk 1]
```diff
@@ -47,12 +47,12 @@ public async Task should_return_too_many_requests_status_code_when_exceding_conf
...
             using var server = new TestServer(webHostBuilder);
 
-            var requests = Enumerable.Range(1, maxActiveRequests)
-                .Select(n => server.CreateRequest($"/healthchecks-api").GetAsync());
+            var requests = Enumerable.Range(1, maxActiveRequests + 2)
+                .Select(_ => server.CreateRequest("/healthchecks-api").GetAsync()).ToList();
 
             var results = await Task.WhenAll(requests).ConfigureAwait(false);
 
-            results.Where(r => r.StatusCode == HttpStatusCode.TooManyRequests).Count().ShouldBe(requests.Count() - maxActiveRequests);
```
Comment (sungam3r) [inline test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs:55]: Before fix `Count()` call executed `requests` enumerable. The same for `Where` leading to unpredicatable results.

File: test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs [hunk 2]
```diff
@@ -1,14 +1,24 @@
...
+
+        public ui_api_request_limiting(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
         [Fact]
         public async Task should_return_too_many_requests_status_code_when_exceding_configured_max_active_requests()
         {
+            _output.WriteLine($"Processors available to should_return_too_many_requests_status_code_when_exceding_configured_max_active_requests: {Environment.ProcessorCount}");
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs:20]: > Processors available to should_return_too_many_requests_status_code_when_exceding_configured_max_active_requests: 2

[the remark this answers is not in the document]
Comment (rob-baldwin) [reply test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs:20]: Correct, standard github runners are very limited and can be a real pain for parallel tests!

https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

File: test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs [hunk 3]
```diff
@@ -47,18 +57,26 @@ public async Task should_return_too_many_requests_status_code_when_exceding_conf
 
             using var server = new TestServer(webHostBuilder);
 
-            var requests = Enumerable.Range(1, maxActiveRequests)
-                .Select(n => server.CreateRequest($"/healthchecks-api").GetAsync());
+            // warmup
+            (await server.CreateRequest("/healthchecks-api").GetAsync().ConfigureAwait(false)).StatusCode.ShouldBe(HttpStatusCode.OK);
+
+            var requests = Enumerable.Range(1, maxActiveRequests + 2)
```
Comment (rob-baldwin) [inline test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs:63]: What about changing this to:

`var requests = Enumerable.Range(1, maxActiveRequests + 2).AsParallel()`

Comment (sungam3r) [reply test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs]: I got it! It's wrong endpoint!

File: src/HealthChecks.UI.InMemory.Storage/HealthChecksUIBuilderExtensions.cs [hunk 4]
```diff
@@ -6,8 +6,12 @@ namespace Microsoft.Extensions.DependencyInjection
     public static class HealthChecksUIBuilderExtensions
     {
         public static HealthChecksUIBuilder AddInMemoryStorage(this HealthChecksUIBuilder builder, Action<DbContextOptionsBuilder>? configureOptions = null, string databaseName = "HealthChecksUI")
+            => builder.AddInMemoryStorage<HealthChecksDb>(configureOptions, databaseName);
+
+        public static HealthChecksUIBuilder AddInMemoryStorage<THealthChecksDb>(this HealthChecksUIBuilder builder, Action<DbContextOptionsBuilder>? configureOptions = null, string databaseName = "HealthChecksUI")
+            where THealthChecksDb : HealthChecksDb
         {
-            builder.Services.AddDbContext<HealthChecksDb>(options =>
+            builder.Services.AddDbContext<HealthChecksDb, THealthChecksDb>(options =>
```
Comment (sungam3r) [inline src/HealthChecks.UI.InMemory.Storage/HealthChecksUIBuilderExtensions.cs:14]: I'll add corresponding methods for other storages in new PR.

Comment (sungam3r) [reply src/HealthChecks.UI.InMemory.Storage/HealthChecksUIBuilderExtensions.cs]: reverted

File: src/HealthChecks.UI/Configuration/Settings.cs [hunk 5]
```diff
@@ -1,9 +1,12 @@
 using HealthChecks.UI.Core;
+using HealthChecks.UI.Data;
+using Microsoft.AspNetCore.Http;
 
 namespace HealthChecks.UI.Configuration
 {
     public class Settings
     {
+        internal Action<List<HealthCheckExecution>>? ConfigureUIApiEndpointResult { get; set; }
```
Comment (sungam3r) [inline src/HealthChecks.UI/Configuration/Settings.cs:9]: internal fo now

File: test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs [hunk 6]
```diff
@@ -17,16 +17,12 @@ public async Task should_return_too_many_requests_status_code_when_exceding_conf
                     services
                         .AddRouting()
                         .AddHealthChecks()
-                        .AddAsyncCheck("Delayed", async () =>
```
Comment (sungam3r) [inline test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs:20]: not applicable here

**Comment by sungam3r** [comment]
Again...
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/actions/runs/5519138963/jobs/10064014608?pr=1887

File: test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs (hunk 3 above)
Comment (sungam3r) [reply test/HealthChecks.UI.Tests/Functional/UIApiRequestLimitingTests.cs:46]: fixed in https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1869

**Comment by sungam3r** [comment]
fixed in https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1869

**Pull request merged**